spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RewriteMergeInto.scala

```diff
@@ -58,53 +61,138 @@ case class RewriteMergeInto(conf: SQLConf) extends Rule[LogicalPlan] with Rewrit
 
   override def apply(plan: LogicalPlan): LogicalPlan = {
     plan resolveOperators {
+      case MergeIntoTable(target: DataSourceV2Relation, source: LogicalPlan, cond, matchedActions, notMatchedActions)
+          if matchedActions.isEmpty =>
+
+        val mergeBuilder = target.table.asMergeable.newMergeBuilder("merge", newWriteInfo(target.schema))
+        val targetTableScan = buildSimpleScanPlan(target.table, target.output, mergeBuilder, cond)
+
+        // when there are no matched actions, use a left anti join to remove any matching rows and rewrite to use
+        // append instead of replace. only unmatched source rows are passed to the merge and actions are all inserts.
+        val joinPlan = Join(source, targetTableScan, LeftAnti, Some(cond), JoinHint.NONE)
```

Comment:
I think we should filter the source before the join if there is a condition in the insert action.

```
...
WHEN NOT MATCHED AND inserCond
  THEN INSERT *
```

Since we run in the optimizer, I am not sure whether Spark will do this automatically for us.

Reply:
Yeah, I was thinking about that, too. I used the merge builder here because that automatically projects `_file` and `_pos`, but it also ignores residuals so we don't get pushdown to data file filters. I'll switch over to a normal scan.

I'll add a filter with the condition as well.

Reply:
Since there may be multiple NOT MATCHED clauses, we cannot really apply the filter.

Reply:
I've updated this so that we apply any filters from the match condition, and to ensure that the predicates can be pushed down to Parquet and ORC.

Reply:
> I think we should filter the source before the join if there is a condition in the insert action
@aokolnychyi Is there a functional issue here ? Currently `MergeIntoExec` would correctly filter the rows ? This is an optimization you r suggesting ?

Reply:
Resolving to make this file readable.